File: scripts/live_lump_guard.py
"""Fail closed for CLI writers aimed at the application's live artifact store."""
from pathlib import Path


def assert_offline_output(path):
    destination = Path(path).resolve()
    live = Path(__file__).resolve().parents[1] / "server" / "lumps"
    live = live.resolve()
    if destination == live or live in destination.parents or destination in live.parents:
        raise RuntimeError(
            "Direct writes to live server/lumps are blocked: build into a temporary "
            "output directory, review the result, then publish through the server's "
            "confirmed admission path. No CLI approval bypass is available.")
    # Temporary directories must not hide links back into the live store.
    # Reject linked outputs rather than overwriting a shared inode.
    if destination.is_dir():
        for item in destination.rglob("*"):
            if item.is_symlink():
                resolved = item.resolve()
                if destination not in resolved.parents or resolved.is_dir():
                    raise RuntimeError("Escaping or directory output links are blocked; use an independent temporary copy.")
                if resolved.is_file() and resolved.stat().st_nlink > 1:
                    raise RuntimeError("Linked output files are blocked; use an independent temporary copy.")
            elif item.is_file() and item.stat().st_nlink > 1:
                raise RuntimeError("Linked output files are blocked; use an independent temporary copy.")
    elif destination.is_file() and destination.stat().st_nlink > 1:
        raise RuntimeError("Linked output files are blocked; use an independent temporary copy.")

File: scripts/live-lump-guard.js
'use strict';
const fs = require('fs');
const path = require('path');

function resolveExisting(target, seen = new Set()) {
    target = path.resolve(target);
    if (seen.has(target)) throw new Error('Cyclic output link is blocked.');
    seen.add(target);
    try {
        if (fs.lstatSync(target).isSymbolicLink())
            return resolveExisting(path.resolve(path.dirname(target), fs.readlinkSync(target)), seen);
    } catch (error) {
        if (error.code !== 'ENOENT') throw error;
    }
    if (fs.existsSync(target)) return fs.realpathSync(target);
    const parent = path.dirname(target);
    return path.join(resolveExisting(parent, seen), path.basename(target));
}

function assertOfflineOutput(target) {
    const destination = resolveExisting(target);
    const live = resolveExisting(path.join(__dirname, '..', 'server', 'lumps'));
    if (destination === live || destination.startsWith(live + path.sep) ||
        live.startsWith(destination + path.sep)) {
        throw new Error('Direct writes to live server/lumps are blocked: build into a temporary output directory, review the result, then publish through the server confirmed admission path. No CLI approval bypass is available.');
    }
    if (fs.existsSync(destination)) {
        const stat = fs.lstatSync(destination);
        if (stat.isDirectory()) {
            for (const name of fs.readdirSync(destination)) {
                const child = path.join(destination, name);
                const childStat = fs.lstatSync(child);
                if (childStat.isSymbolicLink()) {
                    const resolved = resolveExisting(child);
                    if (!resolved.startsWith(destination + path.sep) ||
                        (fs.existsSync(resolved) && fs.statSync(resolved).isDirectory()))
                        throw new Error('Escaping or directory output links are blocked; use an independent temporary copy.');
                }
                assertOfflineOutput(child);
            }
        } else if (stat.isFile() && stat.nlink > 1) {
            throw new Error('Linked output files are blocked; use an independent temporary copy.');
        }
    }
}
module.exports = { assertOfflineOutput };